docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,125 @@
+
+## QoS Functionality
+
+We will introduce two tiers of pods: top-tier pods and best-effort pods. We will not support multiple levels or priorities.
+
+Best-effort pods will be invisible to top-tier pods:
+- **Existence of best-effort pods should not prevent top tier pods from launching**. When deciding whether to launch a top-tier pod, we only look at resource limits of other top-tier pod. Suppose we have a machine with 4GB of memory. Top-tier pod P1 has a 2GB limit, and 4 best-effort pods have a 500MB limit each. Suppose the user launches a new top-tier pod P2 with a 1GB limit. The Kubelet will evict best-effort pods, in unspecified order, to make room for P2. P2 will then be admitted and allocated the requested resources.
+- **Allocation of excess compressible resources (e.g. CPU) to top-tier pods will be the same as if best-effort pods did not exist**. For example if top-tier pod P1 asks for minimum 20% CPU, and we also have best-effort pods B1 and B2 running. Suppose P1 needs more CPU (because of a huge number of requests coming in), then P1 will starve B1, B2 of CPU and use the whole 100% of the CPU (assuming there are no other top tier pods running).
+- **Allocation of non-compressible resources (e.g. memory) to top-tier pods will be the same as if best-effort pods did not exist**. For example consider a system with 4GB of memory, where top-tier pod P1 is using 500MB of memory (with a 1GB limit) and best-effort pods are using the other 3.5GB of memory. The best-effort pods will not prevent P1 from accessing more memory (assuming that P1 remains within its memory limits). So if P1 asks for 100MB more memory, P1 will get the memory it needs, and one of the other pods will be OOM killed. 
+
+Resource guarantees for best-effort pods:
+- Compressible resources
+  - For now, we are only implementing CPU limits.
+  - CPU shares not used by top-tier jobs will be distributed proportionally amongst best-effort pods (based on the amount of CPU the pods requested).
+  - Best effort pods will not be evicted because of CPU constraints, they will simply be starved if CPU resources are not available.
+- Non-compressible resources
+  - For now, we are only implementing memory limits.
+  - We will always launch a best-effort pod, regardless of its memory requirements.
+  - Best effort pods will be evicted if a top tier pod needing the memory requirements comes in, or if some other top-tier pod asks for memory.
+  - Best-effort pods will not be allowed to exceed their memory limit.

```

Comment:
I think we should introduce request, so that the scheduler has some information to work with, even for best-effort pods, and change the default LimitRange to set request rather than limit.